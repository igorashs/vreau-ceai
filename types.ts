import FormidableLib from 'formidable';
import { ValidationError } from 'joi';

export type Product = {
  _id: string;
  name: string;
  category_id: string;
  src: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
};

export type ProductWithCategory = Product & {
  category_id: { name: string };
};

export type ErrorDetail = {
  message: string;
  name?: string;
};

export type ProductName = {
  name: string;
};

export type ProductNameErrorDetail = ErrorDetail & {
  name?: keyof ProductName;
};

export type ProductFields = ProductName & {
  src?: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
  category_id: string;
};

export type Category = {
  _id: string;
  name: string;
};

export type ProductErrorDetail = ErrorDetail & {
  name?: keyof ProductFields;
};

export type ApiResponse = {
  success: boolean;
  message: string;
  errors?: ErrorDetail[];
};

export type User = {
  _id: string;
  name: string;
  email: string;
  isManager: boolean;
};

export type UserEmail = {
  email: string;
};

export type UserEmailErrorDetail = ErrorDetail & {
  name?: keyof UserEmail;
};

export type UserLogin = UserEmail & { password: string };

export type UserSignup = UserEmail & {
  name: string;
  password: string;
  repeat_password: string;
};

export type UserPermissions = { isManager: boolean };

export type LabelMessage = { success: boolean; message: string };

export type UserAuth = {
  _id: string;
  name: string;
  isAdmin: boolean;
  isManager: boolean;
};

export type UserSession = {
  isAuth: boolean;
  user: UserAuth | null;
  needRefresh: boolean;
};

type dataType = 'data' | 'error' | 'field' | 'fileBegin' | 'file' | 'progress';

export declare class Formidable extends FormidableLib {
  emit(name: dataType, data: FormidableLib.EmitData | ValidationError): void;
}

export type CategoryName = {
  name: string;
};

export type CategoryNameErrorDetail = ErrorDetail & {
  name?: keyof CategoryName;
};

export type OrderSubmit = {
  tel: string;
  address: string;
};

export type OrderItem = {
  product_id: string;
  count: number;
};

export type OrderStatus = {
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
};

export type OrderStatusErrorDetail = ErrorDetail & {
  name?: keyof OrderStatus;
};

export type OrderNumber = {
  number: string;
};

export type OrderNumberErrorDetail = ErrorDetail & {
  name?: keyof OrderNumber;
};

export type OrderFields = {
  info: OrderSubmit;
  items: OrderItem[];
};

export type Order = {
  _id: string;
  user: {
    name: string;
    email: string;
  };
  number: string;
  total_price: number;
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
  items: {
    count: number;
    product: { name: string; price: number; quantity: number };
  }[];
  address: string;
  tel: string;
  orderedAt: Date;
  completedAt: Date;
};

export type UserOrders = { orders: Order[]; count: number };

export type ManagementOrders = { orders: Order[]; count: number };
