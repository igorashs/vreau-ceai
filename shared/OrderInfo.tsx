/* eslint-disable camelcase */
import styled from 'styled-components';

const Wrapper = styled.div`
  display: grid;
  gap: var(--baseline);
`;

const List = styled.ul`
  display: grid;
  gap: calc(var(--baseline) / 2);
`;

const Small = styled.small`
  font-weight: 500;
`;

const Price = styled.p`
  font-weight: 500;
`;

const Summary = styled.div`
  display: grid;
  gap: calc(var(--baseline) / 2);
`;

type Product = {
  name: string;
  price: number;
  quantity: number;
};

interface OrderInfoProps {
  order: {
    items: Array<{ product: Product }>;
    total_price: number;
    orderedAt: Date;
    completedAt: Date;
  };
}

export const OrderInfo = ({ order }: OrderInfoProps) => {
  return (
    <Wrapper>
      <List>
        {order.items.map(({ product }) => (
          <li key={product.name}>
            <p>{product.name}</p>
            <Small>
              {product.price}lei({product.quantity}g)
            </Small>
          </li>
        ))}
      </List>

      <Summary>
        <Price>total: {order.total_price}lei</Price>
        <Small>
          {new Date(order.orderedAt).toLocaleDateString()}
          {order.completedAt &&
            `-${new Date(order.completedAt).toLocaleDateString()}`}
        </Small>
      </Summary>
    </Wrapper>
  );
};
