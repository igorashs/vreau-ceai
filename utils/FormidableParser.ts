import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
