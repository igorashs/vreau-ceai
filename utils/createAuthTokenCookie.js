import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
