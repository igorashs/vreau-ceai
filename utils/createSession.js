import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
