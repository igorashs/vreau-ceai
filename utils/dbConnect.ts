import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
