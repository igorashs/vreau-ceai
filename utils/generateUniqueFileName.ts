import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
