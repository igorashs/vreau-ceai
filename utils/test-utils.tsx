import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
