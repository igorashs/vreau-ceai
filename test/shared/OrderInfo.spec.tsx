import { render } from '@/utils/test-utils';
import OrderInfo from '@/shared/OrderInfo';
import { Order } from 'types';

describe('OrderInfo', () => {
  it('renders correctly', () => {
    const order: Order = {
      _id: '1',
      user: {
        name: 'userName',
        email: 'user@u.com',
      },
      number: 'OR123',
      total_price: 500,
      status: 'processing',
      items: [
        {
          count: 1,
          product: { name: 'product1', price: 100, quantity: 50 },
        },
        {
          count: 2,
          product: { name: 'product2', price: 400, quantity: 100 },
        },
      ],
      address: 'add1',
      tel: '061111111',
      orderedAt: new Date('1/1/2021'),
      completedAt: new Date('1/3/2021'),
    };

    const { container, getByText, getAllByRole } = render(
      <OrderInfo order={order} />,
    );

    const listItems = getAllByRole('listitem');

    listItems.forEach((listItem, i) => {
      expect(listItem).toHaveTextContent(order.items[i].product.name);
      expect(listItem).toHaveTextContent(
        order.items[i].product.price.toString(),
      );
      expect(listItem).toHaveTextContent(
        order.items[i].product.quantity.toString(),
      );
    });

    expect(getByText(order.total_price, { exact: false })).toBeVisible();
    expect(
      getByText(order.orderedAt.toLocaleDateString(), { exact: false }),
    ).toBeVisible();

    if (order.completedAt)
      expect(
        getByText(order.completedAt.toLocaleDateString(), { exact: false }),
      ).toBeVisible();

    expect(container.firstChild).toMatchSnapshot();
  });
});
