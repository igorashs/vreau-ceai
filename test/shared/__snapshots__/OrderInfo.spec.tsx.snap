// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderInfo renders correctly 1`] = `
.c0 {
  display: grid;
  gap: var(--baseline);
}

.c1 {
  display: grid;
  gap: calc(var(--baseline) / 2);
}

.c2 {
  font-weight: 500;
}

.c4 {
  font-weight: 500;
}

.c3 {
  display: grid;
  gap: calc(var(--baseline) / 2);
}

<div
  class="c0"
>
  <ul
    class="c1"
  >
    <li>
      <p>
        product1
      </p>
      <small
        class="c2"
      >
        100
        lei(
        50
        g)
      </small>
    </li>
    <li>
      <p>
        product2
      </p>
      <small
        class="c2"
      >
        400
        lei(
        100
        g)
      </small>
    </li>
  </ul>
  <div
    class="c3"
  >
    <p
      class="c4"
    >
      total: 
      500
      lei
    </p>
    <small
      class="c2"
    >
      1/1/2021
      -1/3/2021
    </small>
  </div>
</div>
`;
